### PR TITLE
Add SaveTree, LoadTree

### PR DESCRIPTION
**Pull request by Warchant**

### CODE REVIEW

File: include/veriblock/blockchain/base_block_tree.hpp [hunk 1]
```diff
@@ -239,9 +299,11 @@ struct BaseBlockTree {
...
 
   bool operator==(const BaseBlockTree& o) const {
-    TreeFieldsComparator cmp{};
-    return cmp(blocks_, o.blocks_) && cmp(tips_, o.tips_) &&
-           (activeChain_ == o.activeChain_);
+    CollectionOfPtrComparator cmp{};
+    bool a = cmp(blocks_, o.blocks_);
+    bool b = cmp(tips_, o.tips_);
+    bool c = (activeChain_ == o.activeChain_);
+    return a && b && c;
```
Comment (blackyblack) [inline include/veriblock/blockchain/base_block_tree.hpp:306]: We might avoid unnecessary comparisons using the previous code.

File: include/veriblock/blockchain/block_index.hpp [hunk 2]
```diff
@@ -189,13 +189,14 @@ struct BlockIndex : public Block::addon_t {
     return fromRaw(stream);
   }
 
-  friend bool operator==(const BlockIndex& a, const BlockIndex& b) {
-    return *a.header == *b.header;
+  bool operator==(const BlockIndex& o) const {
+    bool a = *header == *o.header;
+    bool b = status == o.status;
+    bool c = addon_t::operator==(o);
+    return a && b && c;
```
Comment (blackyblack) [inline include/veriblock/blockchain/block_index.hpp:196]: Same here.

File: include/veriblock/blockchain/btc_block_addon.hpp [hunk 3]
```diff
@@ -19,22 +20,25 @@ struct BtcBlockAddon {
   //! reference counter for fork resolution
   uint32_t refCounter = 0;
 
+  bool operator==(const BtcBlockAddon& o) const {
+    bool a = refCounter == o.refCounter;
+    bool b = chainWork == o.chainWork;
+    return a && b;
```
Comment (blackyblack) [inline include/veriblock/blockchain/btc_block_addon.hpp:26]: Same here.

File: include/veriblock/storage/repo_batch_adaptor.hpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    pop_.saveTip(value);
+    return true;
+  }
+  bool writeTip(const BlockIndex<VbkBlock>& value) override {
+    pop_.saveTip(value);
+    return true;
+  }
+  bool writeTip(const BlockIndex<AltBlock>& value) override {
+    pop_.saveTip(value);
+    return true;
```
Comment (blackyblack) [inline include/veriblock/storage/repo_batch_adaptor.hpp:42]: It doesn't look like a batch at all. Is it some kind of a temporary class?

File: include/veriblock/blockchain/base_block_tree.hpp (hunk 1 above)
Comment (Warchant) [reply include/veriblock/blockchain/base_block_tree.hpp:306]: compiler will optimize this away, but at the same time this layout simplifies debugging

File: include/veriblock/storage/repo_batch_adaptor.hpp (hunk 4 above)
Comment (Warchant) [reply include/veriblock/storage/repo_batch_adaptor.hpp:42]: It is not temporary. We use `flush` to actually write on disk, right? If so, it should not matter whether we use batch or no

Comment (blackyblack) [reply include/veriblock/storage/repo_batch_adaptor.hpp:42]: I do not see how the inheritance from the Batch class is being used here. All I see is a collection of PopStorage method wrappers. You could simply use PopStorage methods directly and remove this helper class completely.

**Pull request merged**